File: mefgo/custom/sales_order/sales_order_custom.py
import json
import frappe
from frappe import _

@frappe.whitelist()
def create_mefgo_job_order():
    data = frappe.form_dict.get('data')
    delivery_date = frappe.form_dict.get('delivery_date')
    transaction_date = frappe.form_dict.get('transaction_date')
    name = frappe.form_dict.get('name')

    item = json.loads(data)

    completed = 0
    msg = ''

    if len(item['items']) == 0:
        completed = 1
        msg = _('Kindly select items')

    for i in item['items']:
        i['qty'] = i['pending_qty']
        i['sales_order'] = name
        i['item_description'] = i['description']
        
        comp = frappe.db.sql(f"""SELECT
                                    COALESCE(SUM(mjoi.qty), 0) as qty
                                 FROM `tabMEFGO Job Order Item` mjoi
                                 WHERE mjoi.item_code = %s AND mjoi.docstatus IN %s AND sales_order = %s
                                 LIMIT 1""", (i["item_code"], tuple([0, 1]), name), as_dict=1)

        reqd = frappe.db.get_value('Sales Order Item', i['sales_order_item'], 'qty')
        if len(comp) == 1 and comp[0]['qty'] >= reqd:
            completed = 1
            msg += f"<b>Row {i['idx']}:</b> {i['item_code']} has already been produced in the required quantity <b> ({comp[0]['qty']}) </b>.<br>"

    if completed:
        frappe.throw(msg)


    for i in item['items']:
        
        doc = frappe.get_doc({
        
            "order_date": transaction_date,
            "date_required": delivery_date,
            "ref_no": name,
            "doctype": "MEFGO Job Order",
            "mefgo_job_order_item": [i],
        
        }).insert()
        frappe.db.commit()
        

    frappe.response['message'] = doc.name

File: mefgo/custom/sales_order/sales_order_custom.js
frappe.ui.form.on('Sales Order', {
    refresh(frm) {
       // your code here

       if (cur_frm.doc.docstatus == 1){
         frm.add_custom_button('Job Order', () => {
   
            cur_frm.call({
               method: "erpnext.selling.doctype.sales_order.sales_order.get_work_order_items",
               args: {
                  sales_order: cur_frm.doc.name,
               },
               freeze: true,
               callback: function (r) {
                  if (!r.message) {
                     frappe.msgprint({
                        title: __('Work Order not created'),
                        message: __('No Items with Bill of Materials to Manufacture'),
                        indicator: 'orange'
                     });
                     return;
                  } else {
                     const fields = [{
                        label: 'Items',
                        fieldtype: 'Table',
                        fieldname: 'items',
                        description: __('Select Item and Qty for Production'),
                        fields: [{
                           fieldtype: 'Read Only',
                           fieldname: 'item_code',
                           label: __('Item Code'),
                           in_list_view: 1
                        },
                        {
                           fieldtype: 'Float',
                           fieldname: 'pending_qty',
                           reqd: 1,
                           label: __('Qty'),
                           in_list_view: 1
                        },
                        {
                           fieldtype: 'Data',
                           fieldname: 'sales_order_item',
                           reqd: 1,
                           label: __('Sales Order Item'),
                           hidden: 1
                        }],
                        data: r.message,
                        get_data: () => {
                           return r.message
                        }
                     }]
                     var d = new frappe.ui.Dialog({
                        title: __('Select Items to Manufacture'),
                        fields: fields,
                        primary_action: function () {
                           var data = {
                              items: d.fields_dict.items.grid.get_selected_children()
                           };
                           
                           
                           console.log(data)
                           
                           frappe.call({
                              method: 'mefgo.custom.sales_order.sales_order_custom.create_mefgo_job_order',
                              args: {
                                 data: data,
                                 delivery_date: cur_frm.doc.delivery_date,
                                 transaction_date: cur_frm.doc.transaction_date,
                                 name: cur_frm.doc.name,
   
                              },
                              freeze: true,
                              callback: function (r) {
                                 if (r.message) {
                                    console.log(r)
                                    
                                    frappe.set_route('List', "MEFGO Job Order",{'ref_no':cur_frm.doc.name})
                                    //  frappe.msgprint({
                                    //     message: __('Work Orders Created: {0}', [r.message.map(function (d) {
                                    //       return repl('<a href="/app/work-order/%(name)s">%(name)s</a>', {
                                    //           name: d
                                    //       })
                                    //     }).join(', ')]),
                                    //     indicator: 'green'
                                    //  })
                                 }
                                 d.hide();
                              }
                           });
                           
                           
                        },
                        primary_action_label: __('Create')
                     });
                     d.show();
                  }
               }
            });
   
   
         })
      }

    }
 })